### PR TITLE
Keypair check

### PR DESCRIPTION
**Pull request by dreambeyondorange**
### Description of changes
* Describe *what* you're changing and *why* you're doing these changes.
* Link to impacted open issues.

### Tests
* Describe the automated and/or manual tests executed to validate the patch.
* Describe the added/modified tests.

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/config/cluster_config.py [hunk 1]
```diff
@@ -759,9 +759,6 @@ def __init__(self, key_name: str = None, **kwargs):
         super().__init__(**kwargs)
         self.key_name = Resource.init_param(key_name)
 
-    def _register_validators(self, context: ValidatorContext = None):  # noqa: D102 #pylint: disable=unused-argument
```
Comment (enrico-usai) [inline cli/src/pcluster/config/cluster_config.py:762]: By removing the register_validators from here you're removing the check for LoginNodesSsh. You need to restore it.

File: cli/src/pcluster/validators/ec2_validators.py [hunk 2]
```diff
@@ -171,10 +171,15 @@ class KeyPairValidator(Validator):
     Verify the given key pair is correct.
     """
 
-    def _validate(self, key_name: str):
+    def _validate(self, key_name: str, os: str):
         if key_name:
             try:
-                AWSApi.instance().ec2.describe_key_pair(key_name)
+                key_pair_data = AWSApi.instance().ec2.describe_key_pair(key_name)
+                if os == "ubuntu2204" and key_pair_data["KeyPairs"][0]["KeyType"] == "rsa":
```
Comment (enrico-usai) [inline cli/src/pcluster/validators/ec2_validators.py:178]: rather than parsing the key_pair_data here we could define a `KeyPairInfo` class in the `cli/src/pcluster/aws/aws_resources.py` and encapsulate the internal parsing logic there. 

The `describe_key_pair` method should then return an instance of `KeyPairInfo` and here we can just retrieve the type with: `key_pair.key_type`, this permits to decouple validator/pcluster code from AWS Api internal dictionaries

File: cli/src/pcluster/validators/ec2_validators.py [hunk 3]
```diff
@@ -171,10 +171,15 @@ class KeyPairValidator(Validator):
...
+                    self._add_failure(
+                        "Ubuntu 22.04 does not support RSA keys.  Please generate and use an ed25519 key",
```
Comment (enrico-usai) [inline cli/src/pcluster/validators/ec2_validators.py:180]: nit: extra space before "Please" word.

Comment (enrico-usai) [inline cli/src/pcluster/validators/ec2_validators.py:180]: you can think to generalize this logic, in future AL2023 will have to do the same check, so we could create a dictionary and do something like:

```python
os_key_type = {"ubuntu22": "ed25519"}
if os in os_key_type:
    required_key_type = os_key_type.get(os)
    if key_pair.key_type != required_key_type:
        self._add_failure(f"{os} requires {required_key_type} key. Please generate and use a {required_key_type} key",...
```

File: cli/tests/pcluster/validators/test_ec2_validators.py [hunk 4]
```diff
@@ -332,18 +332,29 @@ def test_instance_type_base_ami_compatible_validator(
...
+            None,
+            AWSClientError(function_name="describe_key_pair", message="does not exist"),
+            "does not exist",
+        ),
+        (GoodKeyPairsDict, "ubuntu2204", None, None),
+        (BadKeyPairsDict, "ubuntu2204", None, "Ubuntu 22.04 does not support RSA keys"),
     ],
 )
-def test_key_pair_validator(mocker, key_pair, side_effect, expected_message):
+def test_key_pair_validator(mocker, key_pair, os, side_effect, expected_message):
```
Comment (enrico-usai) [inline cli/tests/pcluster/validators/test_ec2_validators.py:354]: we should add the tests for the other oses as well, to be sure the logic is not executed.

File: tests/integration-tests/configs/common/common.yaml [hunk 5]
```diff
@@ -394,16 +394,16 @@ scaling:
...
-    dimensions:
-      - regions: ["us-gov-west-1"]
-        instances: {{ common.INSTANCES_DEFAULT_X86 }}
-        oss: ["alinux2"]
-        schedulers: ["awsbatch"]
-      - regions: ["ap-northeast-1"]
-        instances: {{ common.INSTANCES_DEFAULT_ARM }}
-        oss: ["alinux2"]
-        schedulers: ["awsbatch"]
+#  test_awsbatch.py::test_awsbatch:
```
Comment (enrico-usai) [inline tests/integration-tests/configs/common/common.yaml:397]: memo: this should not be part of this patch

**Comment by enrico-usai** [comment]
Memo: before marking it "Ready to review" let's remember to change the title and description

File: cli/src/pcluster/validators/ec2_validators.py (hunk 3 above)
Comment (dreambeyondorange) [reply cli/src/pcluster/validators/ec2_validators.py:180]: Sentences are supposed to have two spaces after a period

File: tests/integration-tests/configs/common/common.yaml (hunk 5 above)
Comment (dreambeyondorange) [reply tests/integration-tests/configs/common/common.yaml:397]: agreed, I was waiting to open this PR until that patch was merged

File: cli/src/pcluster/validators/ec2_validators.py (hunk 3 above)
Comment (enrico-usai) [reply cli/src/pcluster/validators/ec2_validators.py:180]: ok there is an ongoing discussion about this: https://www.instructionalsolutions.com/blog/one-space-vs-two-after-period :D

File: cli/src/pcluster/config/cluster_config.py (hunk 1 above)
Comment (dreambeyondorange) [reply cli/src/pcluster/config/cluster_config.py:762]: This is moving to the BaseClusterConfig for the OS context should should run for any cluster

File: cli/src/pcluster/validators/ec2_validators.py (hunk 3 above)
Comment (dreambeyondorange) [reply cli/src/pcluster/validators/ec2_validators.py:180]: I'm OK with one space!

Comment (dreambeyondorange) [reply cli/src/pcluster/validators/ec2_validators.py:180]: All the keys we generate are ED25519 now, but that doesn't stop users from using different key types like ECDSA, right?  I think we should avoid this complexity for now and we can add the dict when we have more than one case to consider.  And we should probably do the opposite logic, to restrict using RSA rather than checking for the specific use of a single key type

File: cli/src/pcluster/validators/ec2_validators.py (hunk 2 above)
Comment (dreambeyondorange) [reply cli/src/pcluster/validators/ec2_validators.py:178]: Thanks for the idea, will do this!
